### PR TITLE
Break up the message queue by category

### PR DESCRIPTION
**Pull request by hecktest**
Partial fix for #10694, **protocol bump**.

The network channels have been expanded and reorganized into ~~16~~ more categories. Unrelated messages will no longer block one another, hopefully preventing a brief zap of lag from escalating into an unrecoverable lag storm.
A compatibility filter has been added to support protocols 37-39, by mapping the new channels to the three original ones. The legacy mapping is not an exact 1:1 copy of the old one, but instead attempts to balance the load on all three available queues.
In order to avoid having to write more compatibility code later, all 256 possible channel slots have been allocated. I doubt that this will make much difference.

**This PR is Ready for Review** but it needs a lot of testing to make sure nothing blows up.
You might want to squash when merging, there's some silly debugging commits in this branch.

## Further work
Some places in the code still use hardcoded channel IDs. It's not a big deal right now but it would be a good idea to round them up one day and make them use the new enum.

## How to test
Connect to a server running this PR using various supported clients (from protocol 37 to 40).
Check for any instability and edge cases that might arise from the new channel arrangement. Some subtle changes in behavior are expected, but nothing should outright break.
Compare the network performance to a server without this PR, especially in low bandwidth conditions.

### CODE REVIEW

**Comment by hecktest** [comment]
Any idea what's tripping up testConnectSendReceive on that particular check?

File: src/network/serveropcodes.cpp [hunk 1]
```diff
@@ -155,71 +152,95 @@ const ClientCommandFactory clientCommandFactoryTable[TOCLIENT_NUM_MSG_TYPES] =
 	null_command_factory, // 0x1D
 	null_command_factory, // 0x1E
 	null_command_factory, // 0x1F
-	{ "TOCLIENT_BLOCKDATA",                2, true }, // 0x20
-	{ "TOCLIENT_ADDNODE",                  0, true }, // 0x21
-	{ "TOCLIENT_REMOVENODE",               0, true }, // 0x22
+	{ "TOCLIENT_BLOCKDATA",                MTSCMC_MAP, true }, // 0x20
```
Comment (lhofhansl) [inline src/network/serveropcodes.cpp:155]: So these three were using 2 channels before, I think we should keep them on separate channels, if only to indicate that there is no ordering assumption between theses.

Comment (lhofhansl) [reply src/network/serveropcodes.cpp:155]: Perhaps another channel MT_SCMC_MAPCHANGE.

File: src/network/serveropcodes.cpp [hunk 2]
```diff
@@ -155,71 +152,95 @@ const ClientCommandFactory clientCommandFactoryTable[TOCLIENT_NUM_MSG_TYPES] =
...
+	{ "TOCLIENT_OVERRIDE_DAY_NIGHT_RATIO", MTSCMC_ENVIRONMENT, true }, // 0x50
+	{ "TOCLIENT_LOCAL_PLAYER_ANIMATIONS",  MTSCMC_ENTITY, true }, // 0x51
+	{ "TOCLIENT_EYE_OFFSET",               MTSCMC_CAMERA, true }, // 0x52
+	{ "TOCLIENT_DELETE_PARTICLESPAWNER",   MTSCMC_PARTICLE, true }, // 0x53
+	{ "TOCLIENT_CLOUD_PARAMS",             MTSCMC_ENVIRONMENT, true }, // 0x54
+	{ "TOCLIENT_FADE_SOUND",               MTSCMC_AUDIO, true }, // 0x55
+	{ "TOCLIENT_UPDATE_PLAYER_LIST",       MTSCMC_AUTH, true }, // 0x56
+	{ "TOCLIENT_MODCHANNEL_MSG",           MTSCMC_MODCHANNEL, true }, // 0x57
+	{ "TOCLIENT_MODCHANNEL_SIGNAL",        MTSCMC_MODCHANNEL, true }, // 0x58
+	{ "TOCLIENT_NODEMETA_CHANGED",         MTSCMC_MAP, true }, // 0x59
```
Comment (lhofhansl) [inline src/network/serveropcodes.cpp:212]: Same here. I wouldn't put these on the MAP channel. This could be on the same as ADDNODE/REMOVENODE or even a separate one.

File: src/network/serveropcodes.cpp (hunk 1 above)
Comment (hecktest) [reply src/network/serveropcodes.cpp:155]: Makes sense. And hey, it's pretty easy to change this stuff now.

**Comment by hecktest** [comment]
I've noticed some issues with the server sometimes failing to send the correct time of day on login. Maybe some special treatment is needed for init packets until the player is spawned proper.

**Comment by KaylebJay** [comment]
Should be another blocker for 5.4! Lots of clients get into this mess currently, doesn't help promote minetest at all...

**Comment by hecktest** [comment]
@KaylebJay
Please test this branch if you want it in quickly. The more eyes are on it, the faster we can find all the edge cases.

**Comment by kilbith** [comment]
It’d get more testing if it was merged. Applying a patch is more painful than pulling.

**Comment by Jipok** [comment]
> It’d get more testing if it was merged. Applying a patch is more painful than pulling.

What's the problem?
```
git remote add hecktest https://github.com/hecktest/minetest
git checkout hecktest/queue-death
```

**Comment by SmallJoker** [comment]
>Lots of clients get into this mess currently, doesn't help promote minetest at all...

Race-conditions, and resulting client stalls doesn't help to promote Minetest at all. Like every other potentially error-prone PR, this has to be tested as well.

> Applying a patch is more painful than pulling.

FYI: this can be automated. [apply_pr script](https://gist.github.com/SmallJoker/4d411b958295d9e694cd9aa5984563db/) for `~/.local/bin`

**Comment by lhofhansl** [comment]
> I've noticed some issues with the server sometimes failing to send the correct time of day on login. Maybe some special treatment is needed for init packets until the player is spawned proper.

Might be a previous ordering enforced by the using fewer channels.

@hecktest What's the best way to the improvement I should see from this? Many entities?

**Comment by hecktest** [comment]
@lhofhansl 
Any situation where bandwidth from the server to the client is tight. The PR should tolerate a lower bandwidth before the server becomes unplayable.

**Comment by lhofhansl** [comment]
I definitely like this approach! By putting messages on the same channel we also document (if you will) that there is an expected ordering between them (and likewise that there is not between channels).

I'll try it some more.

**Comment by KaylebJay** [comment]
> @KaylebJay
> Please test this branch if you want it in quickly. The more eyes are on it, the faster we can find all the edge cases.

Sorry, didn't see this notification & one of my friends just alerted me to it. I would love to test it, will do that when I find some time.

**Comment by appgurueu** [comment]
This most likely breaks `minetest.dynamic_add_media` (`Server::dynamicAddMedia`) if the media is used immediately after the call due to possible race conditions. A comment states:
> The network layer only guarantees ordered delivery inside a channel.
> Since the very next packet could be one that uses the media, we have
> to push the media over ALL channels to ensure it is processed before
> it is used. In practice this means we have to send it twice:
> - channel 1 (HUD)
> - channel 0 (everything else: e.g. play_sound, object messages)

You would essentially have to rewrite that method, first sending the media packet, and *then* sending packets on all other channels which might require the media packet (such as sound, object properties etc.) instructing the client to *await* it's arrival.

**Comment by hecktest** [comment]
Well that's a problem, think about it: if it's not safe to use any API during a media push, it becomes a synchronous process.
Media is usually big, so we're potentially blocking all clients for several seconds. At this point, this is barely any better than just restarting the server to update the assets! All to support realtime grinch skin updates. Sounds like this misfeature just wasn't very well thought out, which should've been obvious when the apparent workaround became "send it twice".

But let's try to salvage this - start requiring a fallback static asset for all dynamic media. That way, if due to a race condition such an asset is used but not yet received, the result is merely annoying and not catastrophic - and we can use some sort of dirty flag to update and reload the asset as soon as the real thing arrives.

Or, there's one more solution, someone suggested it to me just now - accept a usage tag for each asset, which would internally correspond to a network channel. For example, audio assets would belong on the audio channel and play_sound would wait for them. Textures and meshes could be tagged for use with objects, forms, fx or hud - and would be sent on those respective channels. Maybe even allow multiple channels per asset, codifying the multi-send workaround in API, and allowing someone to choose if they want to resort to that.

**Comment by appgurueu** [comment]
> Media is usually big, so we're potentially blocking all clients for several seconds. At this point, this is barely any better than just restarting the server to update the assets! All to support realtime grinch skin updates. Sounds like this misfeature just wasn't very well thought out, which should've been obvious when the apparent workaround became "send it twice".

"Send it twice" wouldn't be an issue if it wasn't the entire file being sent twice. Also, it doesn't have to block the server - it only has to block the clients.

Unfortunately, you have to support this "misfeature" if you want this in 5.4...

**Comment by hecktest** [comment]
If you push media while the game is running, all clients will need to receive it. Blocking all clients is even worse than pausing the server, because mobs, lava, weather, etc. can still kill them. Instead of sending the data on demand, the data is broadcast to all clients in one big burst, possibly causing a huge lag storm and making everyone reconnect anyway. And it _is_ being sent twice and that wasn't my brilliant idea either.

When that crap was getting merged, I cautioned about its dubious utility, and inferiority to something like embedding base64 blobs in the name.
Normally I would agree that a PR must support an existing feature, but in this case, _the feature was never finished_. The problem of API safety was never solved, instead an ogre solution was patched into it. (How did that _ever_ get approved?)
And now I actually cannot support dynamic media without some sort of concession like a channel parameter, or malicious compliance in the form of sending the file over all 17 channels. (If using two was okay, why not 17?)

Let me think... maybe "media is pushed to everyone and not on demand" is the problem? Maybe a media send should occur just-in-time right before a packet referencing said media, on the same channel. That way, it'll be safe, it'll actually make some damn sense and be truly dynamic, saving bandwidth, it will largely avoid the double-send but still do it in the very rare case a file is used in two contexts simultaneously, and it'll even put a foot in the door for streamed assets like audio, or deferring some assets to reduce load times.
You know, dynamic media can actually be useful beyond grinch skins, but the current implementation is very poorly thought out.

**Comment by sfan5** [comment]
Hm yeah, changes to the packet channels do indeed break the current implementation of `dynamicAddMedia`.
I had plans to improve it and make it async so the 'send it twice' hack would no longer be required but I didn't expect it to be an issue so soon.

> Start requiring a fallback static asset for all dynamic media. That way, if due to a race condition such an asset is used but not yet received, the result is merely annoying and not catastrophic - and we can use some sort of dirty flag to update and reload the asset as soon as the real thing arrives.

This is easily doable for sounds but I see some caveats (or rather extra work) for textures and models.
AFAIK textures can only be replaced while keeping `ITexture*` pointer valid if their dimensions match, so for other cases some poll for reload mechanism like you mentioned would have to be implemented. Models are created from their string representation on-demand and resulting objects not kept track of, so objects would also need to poll for mesh reloading.

> Accept a usage tag for each asset, which would internally correspond to a network channel. For example, audio assets would belong on the audio channel and play_sound would wait for them.

I considered this solution before but came to the conclusion that it's not a good idea because it exposes implementation details of the network layer to mods.
Had I added asset tags to differ "hud" vs everything else we'd still have the same problem now. Any future reworks of network channels would also run risk of the current set of asset tags not being precise enough to correctly attribute usage to network channels.

> Maybe a media send should occur just-in-time right before a packet referencing said media, on the same channel. That way, it'll be safe, it'll actually make some damn sense and be truly dynamic, [...]

I also briefly thought about this approach but the reason I haven't done this is that it requires extensive and invasive changes to track where a string put into some API ends up on the other side. It'd be a very elegant approach though.

Also, I don't want to revive the implementation discussion but I will say that I do not consider embedded base64 a superior solution. 

<hr />

> Unfortunately, you have to support this "misfeature" if you want this in 5.4...

This would ordinarily be the case, but due to the circumstances and small timeframe `dynamic_add_media` has existed I support breaking backwards compatibility and reimplementing it in a way that is compatible with this PR. (Though you'll have to hear other coredevs opinions' on this.)

The solution I'm thinking of is to make dynamic_add_media include a callback so that mods are informed (per-client) when the media is ready to use. This would solve any dependency on network channels and also allow the client to use existing remote_media servers to grab it.
Might also be worth to take a look at what people have built using `dynamic_add_media` so that no usecase is left behind.

**Comment by hecktest** [comment]
Found a bit of a snag; forms can now arrive before the inventories they reference. This isn't a problem in practice, because as soon as the inventory arrives, it is displayed in the form - but this annoying warning is first shown. What do we do with it? Only things like dynamically allocated lists are truly affected, maybe it could be suppressed with something like a `warn[false]` element?
invlist

**Comment by lhofhansl** [comment]
I think all message that have implicit ordering dependencies should use the same channel. Can we do this here, or would this break the purpose of this PR?

**Comment by hecktest** [comment]
>I think all message that have implicit ordering dependencies should use the same channel. Can we do this here, or would this break the purpose of this PR?

In this particular case I'd much rather make forms more robust against this condition. In fact they deal with it just fine other than the message, I don't see much of a problem with just suppressing this warning in some way. This is about the only problem that has emerged so far.

Every channel merge increases the chances of a queue getting clogged. Not all forms contain an inventory, should they wait for inventory updates? Probably not.

**Comment by nauta-turbidus** [comment]
> "Send it twice" wouldn't be an issue if it wasn't the entire file being sent twice.

"Send it twice" might be an issue in any case if the second send can again be treated the same way, and sent twice, ad infinitum.

Really we must be extra careful with any protocol changes to prevent loops of resending packets so much the server DDoSes itself (some proprietary games do that).
